### PR TITLE
Fix: POST v2/subscriptions in notification/attrs the subsc is created with attribute names equals (#2101)

### PR DESCRIPTION
**Pull request by Anjali-NEC**
Fix issue #2101

### CODE REVIEW

File: src/lib/jsonParseV2/parseStringVector.cpp [hunk 1]
```diff
@@ -69,6 +69,17 @@ bool parseStringVector
 
     std::string value = iter->GetString();
 
+   for (rapidjson::Value::ConstValueIterator iter_new = iter+1; iter_new != jsonVector.End(); ++iter_new)
```
Comment (fgalan) [inline src/lib/jsonParseV2/parseStringVector.cpp:72]: I think the addition of this code comment can be useful:

```suggestion
   // We start the loop in next position of the vector (note the iter+1), so check that in the elements "to the right" there aren't duplicates
   for (rapidjson::Value::ConstValueIterator iter_new = iter+1; iter_new != jsonVector.End(); ++iter_new)
```

File: CHANGES_NEXT_RELEASE [hunk 2]
```diff
@@ -0,0 +1 @@
+Fix: POST v2/subscriptions in notification/attrs the subsc is created with attribute names equals (#2101)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:1]:
```suggestion
Fix: avoid duplicated attributes in arrays in POST v2/subscriptions (#2101)
```

File: src/lib/jsonParseV2/parseStringVector.cpp (hunk 1 above)
Comment (Anjali-NEC) [reply src/lib/jsonParseV2/parseStringVector.cpp:73]: Fixed in https://github.com/telefonicaid/fiware-orion/pull/4025/commits/ec093717221e85d86e02710201cb026c47f49dbd

**Review by fgalan — APPROVED** [review]
~LGTM~. Nice job!

**Comment by fgalan** [comment]
> LGTM. Nice job!

I went too fast.. sorry :)

In fact, extra modifications are needed. Note the following failiing tests:

```
----------- Failing tests ------------------
  o  3311_duplicate_attrs_in_filter/duplicate_attrs_in_filter_sub.test
  o  3311_duplicate_metadata_in_filter/duplicate_metadata_in_filter_sub.test
```

These tests were introduced due to issue #3311 to check that attribute duplication in subscription were not causing duplicating attributes in notifications. However, given the fix you have done this checking is no longer needed.

Thus, actions to take:

* Remove test/functionalTest/cases/3311_duplicate_attrs_in_filter/duplicate_attrs_in_filter_sub.test
* Remove test/functionalTest/cases/3311_duplicate_metadata_in_filter/duplicate_metadata_in_filter.test
* Modify the .test you have included in this PR to include also the checking for duplicated elements in `metadata` array

**Comment by fgalan** [comment]
Note also the lint fail:

```
0004 errors of category 'Weird number of spaces at line-start'
...
src/lib/jsonParseV2/parseStringVector.cpp:72:  Weird number of spaces at line-start.  Are you using a 2-space indent?  [whitespace/indent] [3]
src/lib/jsonParseV2/parseStringVector.cpp:73:  Weird number of spaces at line-start.  Are you using a 2-space indent?  [whitespace/indent] [3]
src/lib/jsonParseV2/parseStringVector.cpp:74:  Weird number of spaces at line-start.  Are you using a 2-space indent?  [whitespace/indent] [3]
src/lib/jsonParseV2/parseStringVector.cpp:82:  Weird number of spaces at line-start.  Are you using a 2-space indent?  [whitespace/indent] [3]
```

**Comment by Anjali-NEC** [comment]
> Note also the lint fail:
> 
> ```
> 0004 errors of category 'Weird number of spaces at line-start'
> ...
> src/lib/jsonParseV2/parseStringVector.cpp:72:  Weird number of spaces at line-start.  Are you using a 2-space indent?  [whitespace/indent] [3]
> src/lib/jsonParseV2/parseStringVector.cpp:73:  Weird number of spaces at line-start.  Are you using a 2-space indent?  [whitespace/indent] [3]
> src/lib/jsonParseV2/parseStringVector.cpp:74:  Weird number of spaces at line-start.  Are you using a 2-space indent?  [whitespace/indent] [3]
> src/lib/jsonParseV2/parseStringVector.cpp:82:  Weird number of spaces at line-start.  Are you using a 2-space indent?  [whitespace/indent] [3]
> ```

Fixed in https://github.com/telefonicaid/fiware-orion/pull/4025/commits/cb498641c2869966b0317ba5421b4931aeb67854

**Comment by Anjali-NEC** [comment]
> Thus, actions to take:
> 
> * Remove test/functionalTest/cases/3311_duplicate_attrs_in_filter/duplicate_attrs_in_filter_sub.test
> * Remove test/functionalTest/cases/3311_duplicate_metadata_in_filter/duplicate_metadata_in_filter.test
> * Modify the .test you have included in this PR to include also the checking for duplicated elements in `metadata` array

Fixed as per comment in https://github.com/telefonicaid/fiware-orion/pull/4025/commits/73b8b93187df09fe9c9bd2bc425fc82f5c20dc83

**Comment by Anjali-NEC** [comment]
@fgalan sir,

Done modifications in my latest commit. Now all test cases are pass. if it is ok then please merge it into master.

File: test/functionalTest/cases/2101_Subscription_is_not_created_with_attribute_names_equals/Subscription_is_not_created_with_attribute_names_equals.test [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+07. POST /v2/subscriptions, to create subscription with N attributes with the same name in metadata and see error
+=================================================================================================================
+HTTP/1.1 400 Bad Request
+Content-Length: 73
+Content-Type: application/json
+Fiware-Correlator: REGEX([0-9a-f\-]{36})
+Date: REGEX(.*)
+
+{
+    "description": "duplicated attribute /temperature/",
```
Comment (fgalan) [inline test/functionalTest/cases/2101_Subscription_is_not_created_with_attribute_names_equals/Subscription_is_not_created_with_attribute_names_equals.test:391]: In this case the description is a bit missleading, as the problem is not in attributes but in metadata. Maybe it should be changes to something more neutral, like:

```suggestion
    "description": "duplicated item in array /temperature/",
```

File: test/functionalTest/cases/2101_Subscription_is_not_created_with_attribute_names_equals/Subscription_is_not_created_with_attribute_names_equals.test [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+
+#
+# 01. POST /v2/subscriptions, to create subscription with N attributes with the same name in attrs and see error
+# 02. POST /v2/subscriptions, to create subscription with N attributes with the same name in condition and see error
+# 03. POST /v2/subscriptions, to create subscription with N attributes with the same name in exceptAttrs and see error
+# 04. POST /v2/subscriptions, to create subscription with N attributes with different name in attrs and see no error
+# 05. POST /v2/subscriptions, to create subscription with N attributes with different name in condition and see no error
+# 06. POST /v2/subscriptions, to create subscription with N attributes with different name in exceptAttrs and see no error
+# 07. POST /v2/subscriptions, to create subscription with N attributes with the same name in metadata and see error
+# 08. POST /v2/subscriptions, to create subscription with N attributes with different name in metadata and see no error
```
Comment (fgalan) [inline test/functionalTest/cases/2101_Subscription_is_not_created_with_attribute_names_equals/Subscription_is_not_created_with_attribute_names_equals.test:40]: I'd suggest the following ordering in steps (all errors first, all ok second):

```suggestion
# 01. POST /v2/subscriptions, to create subscription with N attributes with the same name in attrs and see error
# 02. POST /v2/subscriptions, to create subscription with N attributes with the same name in condition and see error
# 03. POST /v2/subscriptions, to create subscription with N attributes with the same name in exceptAttrs and see error
# 04. POST /v2/subscriptions, to create subscription with N attributes with the same name in metadata and see error
# 05. POST /v2/subscriptions, to create subscription with N attributes with different name in attrs and see no error
# 06. POST /v2/subscriptions, to create subscription with N attributes with different name in condition and see no error
# 07. POST /v2/subscriptions, to create subscription with N attributes with different name in exceptAttrs and see no error
# 08. POST /v2/subscriptions, to create subscription with N attributes with different name in metadata and see no error
```

(Of course, not only in this commented preamble but also the actual ordering of steps)

File: CHANGES_NEXT_RELEASE [hunk 5]
```diff
@@ -0,0 +1 @@
+Fix: avoid duplicated attributes in arrays in POST v2/subscriptions (#2101)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:1]:
```suggestion
Fix: avoid duplicated attributes and metadata in arrays in POST /v2/subscriptions (#2101)
```

**Comment by fgalan** [comment]
@Anjali-NEC I understand the fix you have done also covers the subscriptions updates with `PATCH /v2/subscriptions/{subId}`, right?

In that case, I'd suggest to extend the test coverage also to the `PATCH /v2/subscriptions/{subId}` operation. It would be just a matter of using a similar .test to the one already (same steps, with a previous one to create the subscription to update). It should be easy to do.

File: CHANGES_NEXT_RELEASE (hunk 5 above)
Comment (Anjali-NEC) [reply CHANGES_NEXT_RELEASE]: Fixed in https://github.com/telefonicaid/fiware-orion/pull/4025/commits/81e67f5fb86e4e61d4879e6ee8924b88fdd5d5cb

File: test/functionalTest/cases/2101_Subscription_is_not_created_with_attribute_names_equals/Subscription_is_not_created_with_attribute_names_equals.test (hunk 4 above)
Comment (Anjali-NEC) [reply test/functionalTest/cases/2101_Subscription_is_not_created_with_attribute_names_equals/Subscription_is_not_created_with_attribute_names_equals.test]: Fixed in https://github.com/telefonicaid/fiware-orion/pull/4025/commits/81e67f5fb86e4e61d4879e6ee8924b88fdd5d5cb

File: test/functionalTest/cases/2101_Subscription_is_not_created_with_attribute_names_equals/Subscription_is_not_created_with_attribute_names_equals.test (hunk 3 above)
Comment (Anjali-NEC) [reply test/functionalTest/cases/2101_Subscription_is_not_created_with_attribute_names_equals/Subscription_is_not_created_with_attribute_names_equals.test]: Fixed in https://github.com/telefonicaid/fiware-orion/pull/4025/commits/81e67f5fb86e4e61d4879e6ee8924b88fdd5d5cb

**Comment by Anjali-NEC** [comment]
> In that case, I'd suggest to extend the test coverage also to the `PATCH /v2/subscriptions/{subId}` operation. It would be just a matter of using a similar .test to the one already in this PR (same steps, with a previous one to create the subscription to update). It should be easy to do.

Added test case for `PATCH /v2/subscriptions/{subId}` in commit https://github.com/telefonicaid/fiware-orion/pull/4025/commits/81e67f5fb86e4e61d4879e6ee8924b88fdd5d5cb

File: test/functionalTest/cases/3311_duplicate_metadata_in_filter/duplicate_metadata_in_filter.test [hunk 6]
```diff
@@ -1,179 +0,0 @@
-# Copyright 2018 Telefonica Investigacion y Desarrollo, S.A.U
```
Comment (fgalan) [inline test/functionalTest/cases/3311_duplicate_metadata_in_filter/duplicate_metadata_in_filter.test:1]: This .test is not related with subscriptions, so it should not be deleted in this PR.

(Maybe is has been removed due to my comment https://github.com/telefonicaid/fiware-orion/pull/4025#issuecomment-995033495. Sorry I actually mean duplicate_metadata_in_filter_sub.test...)

File: CHANGES_NEXT_RELEASE [hunk 7]
```diff
@@ -1,2 +1,3 @@
+- Fix: avoid duplicated attributes and metadata in arrays in POST /v2/subscriptions (#2101)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:1]:
```suggestion
- Fix: avoid duplicated attributes and metadata in arrays in POST /v2/subscriptions and PATCH /v2/subscription/{subId} (#2101)
```

**Comment by fgalan** [comment]
I think the PR is almost ready. Only a couple of new small comments are pending.

File: CHANGES_NEXT_RELEASE (hunk 7 above)
Comment (Anjali-NEC) [reply CHANGES_NEXT_RELEASE]: Fixed in https://github.com/telefonicaid/fiware-orion/pull/4025/commits/dc2c8bbc49fc88cb4a790524abd1a86a01dd1387

File: test/functionalTest/cases/3311_duplicate_metadata_in_filter/duplicate_metadata_in_filter.test (hunk 6 above)
Comment (Anjali-NEC) [reply test/functionalTest/cases/3311_duplicate_metadata_in_filter/duplicate_metadata_in_filter.test:1]: Added test case in https://github.com/telefonicaid/fiware-orion/pull/4025/commits/7922dec240c6daecc35fcb2b4abec7fc1de3f1f0

Comment (fgalan) [reply test/functionalTest/cases/3311_duplicate_metadata_in_filter/duplicate_metadata_in_filter.test:1]: test/functionalTest/cases/3311_duplicate_metadata_in_filter/duplicate_metadata_in_filter.test stills appearing as deleted in "Files changed" tab.

imagen

Please check.

Comment (Anjali-NEC) [reply test/functionalTest/cases/3311_duplicate_metadata_in_filter/duplicate_metadata_in_filter.test]: I have checked.  By mistake I have added another file sorry for this. Now I have added test/functionalTest/cases/3311_duplicate_metadata_in_filter/duplicate_metadata_in_filter.test in my last commit https://github.com/telefonicaid/fiware-orion/pull/4025/commits/3c4540e36aa51cd8c73feea618af899cfe3d43b6

Please check "Files changed" tab.

**Comment by Anjali-NEC** [comment]
@fgalan, Test cases are failing after my modifications. I don't know why. do you have any idea about it?

**Comment by fgalan** [comment]
I see this error:

```
In file included from /opt/fiware-orion/src/lib/rest/ConnectionInfo.h:42,
                 from /opt/fiware-orion/src/lib/serviceRoutines/versionTreat.h:31,
                 from /opt/fiware-orion/src/lib/common/globals.cpp:38:
/opt/fiware-orion/src/lib/rest/mhd.h:31:10: fatal error: microhttpd.h: No such file or directory
 #include <microhttpd.h>
```

It seems not related with the PR at all. Maybe something has broken de CI docker image we are using to run the tests. I'll have a look.

**Comment by fgalan** [comment]
At last CI build image (https://github.com/telefonicaid/fiware-orion/runs/4582544311?check_suite_focus=true, 20 dic 2021 14:08 CET):

```
curl: (7) Failed to connect to ftp.gnu.org port 80: Connection timed out
```

**Comment by fgalan** [comment]
> At last CI build image (https://github.com/telefonicaid/fiware-orion/runs/4582544311?check_suite_focus=true, 20 dic 2021 14:08 CET):
> 
> ```
> curl: (7) Failed to connect to ftp.gnu.org port 80: Connection timed out
> ```

It seem it was a temporal fail. Now tests have passed.

**Review by fgalan — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**